### PR TITLE
Upgrade (not Kotlin) and other dependencies

### PR DESCRIPTION
**Pull request by SalvatoreT**
To get a list of dependencies that could be updated, I used the versions plugin.

    gradle dependencyUpdates -Drevision=release

There were a number of libraries that I couldn't get upgraded quickly (dokka and ktlint among them), but hopefully this helps move thing forward.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -105,7 +106,7 @@ subprojects {
 }
 
 apply(from = rootProject.file(".buildscript/binary-validation.gradle"))
-
+apply(plugin = "com.github.ben-manes.versions")
```
Comment (zach-klippenstein) [inline build.gradle.kts:109]: Why was this added?

File: samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/board/Parser.kt [hunk 2]
```diff
@@ -106,7 +103,7 @@ private fun BufferedSource.readHeader(): String? = buildString {
   while (true) {
     val line = readUtf8Line() ?: throw IllegalArgumentException("Expected --- but found EOF.")
     if (line == YAML_DELIMITER) return@buildString
-    appendln(line)
+    appendLine(line)
```
Comment (zach-klippenstein) [inline samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/board/Parser.kt:106]: We can't use the stdlib from kotlin 1.4 yet, since we can't use kotlin 1.4 internally.

File: versions.properties [hunk 3]
```diff
@@ -3,46 +3,47 @@
...
 version.io.reactivex.rxjava2..rxandroid=2.1.1
 version.io.reactivex.rxjava2..rxjava=2.2.19
 version.junit.junit=4.13
-version.kotlin=1.3.72
-version.kotlinx.coroutines=1.3.7
-version.kotlinx.serialization=0.20.0
-version.me.champeau.gradle..jmh-gradle-plugin=0.5.0
-version.moshi=1.9.3
+version.kotlin=1.4.10
+version.kotlinx.coroutines=1.3.9
```
Comment (zach-klippenstein) [inline versions.properties:37]: Can't upgrade coroutines yet because we can't require the 1.4 stdlib.

File: versions.properties [hunk 4]
```diff
@@ -3,46 +3,47 @@
...
 version.io.mockk..mockk=1.10.0
 version.io.reactivex.rxjava2..rxandroid=2.1.1
 version.io.reactivex.rxjava2..rxjava=2.2.19
 version.junit.junit=4.13
-version.kotlin=1.3.72
-version.kotlinx.coroutines=1.3.7
-version.kotlinx.serialization=0.20.0
-version.me.champeau.gradle..jmh-gradle-plugin=0.5.0
-version.moshi=1.9.3
+version.kotlin=1.4.10
```
Comment (zach-klippenstein) [inline versions.properties:36]: We can build with 1.4, but we need to restrict to the 1.3.x stdlib.

**Review by zach-klippenstein — CHANGES_REQUESTED** [review]
Unfortunately we can't really use 1.4 yet.

File: build.gradle.kts (hunk 1 above)
Comment (SalvatoreT) [reply build.gradle.kts:109]: This allows us to see which libraries need updating with the following command:

```
gradle dependencyUpdates -Drevision=release
```

File: samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/board/Parser.kt (hunk 2 above)
Comment (SalvatoreT) [reply samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/board/Parser.kt:106]: ☹️

File: versions.properties (hunk 3 above)
Comment (SalvatoreT) [reply versions.properties:37]: ☹️

File: build.gradle.kts (hunk 1 above)
Comment (zach-klippenstein) [reply build.gradle.kts:109]: We're already using one plugin to manage dependencies, I'm very reluctant to add a second.

**Comment by SalvatoreT** [comment]
updated to remove Kotlin 1.4 and libraries that require Kotlin 1.4

File: build.gradle.kts [hunk 5]
```diff
@@ -105,7 +105,6 @@ subprojects {
 }
 
 apply(from = rootProject.file(".buildscript/binary-validation.gradle"))
-
```
Comment (zach-klippenstein) [inline build.gradle.kts:108]: This blank line was intentional, since the below lines are unrelated.

**Review by zach-klippenstein — CHANGES_REQUESTED** [review]
One last (accidental?) change that needs to be reverted, but then we can merge this.

**Comment by SalvatoreT** [comment]
Updated!

**Comment by zach-klippenstein** [comment]
LG, that instrumentation test failure was an infra issue but I want to see it pass just to make sure, then I'll merge.
